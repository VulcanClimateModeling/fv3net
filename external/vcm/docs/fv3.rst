Working with FV3 outputs
========================

FV3GFS restart files
--------------------

I/O
~~~

Restart files can be loaded from disk with the functions

.. autofunction:: vcm.fv3_restarts.open_restarts


Coarse-graining
~~~~~~~~~~~~~~~

VCM Tools provides two high-level methods for coarse-graining a complete set of
restart files produced from running FV3GFS.  The two methods perform the
coarse-graining in different ways.

.. autofunction:: vcm.cubedsphere.coarsen_restarts_on_sigma
.. autofunction:: vcm.cubedsphere.coarsen_restarts_on_pressure

Parsing FV3 logs
----------------

Use the following routines to parse statistical information from the console outputs of an FV3 run:

.. autofunction:: vcm.fv3.logs.loads


Standardizing FV3 outputs
-------------------------

The diagnostics output by FV3GFS have verbose, and sometimes difficult to
understand dimension names.  The functions here provide convenient ways for
renaming dimensions to and from standard names, as well as standardizing time
coordinate information:

.. automodule:: vcm.fv3.metadata
   :members:
   :undoc-members:
   :show-inheritance:
