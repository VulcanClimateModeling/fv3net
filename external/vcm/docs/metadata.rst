Renaming FV3GFS dimensions
==========================

The diagnostics output by FV3GFS have verbose, and sometimes difficult to
understand dimension names.  The functions here provide convenient ways for
renaming dimensions to and from standard names.

.. automodule:: vcm.metadata
   :members:
   :undoc-members:
   :show-inheritance:
