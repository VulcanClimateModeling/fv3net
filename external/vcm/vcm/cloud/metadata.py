import dataclasses
import json
from typing import Mapping, List


@dataclasses.dataclass
class StepMetadata:
    job_type: str
    commit: str
    url: str
    dependencies: Mapping[str, str] = dataclasses.field(default_factory=dict)
    args: List[str] = dataclasses.field(default_factory=list)

    def print_json(self):
        print(json.dumps(dataclasses.asdict(self)))
