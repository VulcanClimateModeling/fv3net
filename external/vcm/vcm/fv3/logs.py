#!/usr/bin/env python
import re
from typing import Mapping, List, Tuple
from collections import defaultdict
from datetime import datetime
from toolz import curry
import dataclasses


__all__ = ["loads"]


def parse_date_stats_block(line):
    tokens = line.split()
    args = [int(tok) for tok in tokens]
    return datetime(*args)


@curry
def parse_pattern(pattern, line):
    water_vapor_regex = re.compile(pattern + r" *= (.*)")
    match = water_vapor_regex.search(line)
    if match:
        return float(match.group(1).strip())
    else:
        return None


def set_initial_date(dates):
    if dates[0] is None:
        initial_date = dates[1] - (dates[2] - dates[1])
        return [initial_date] + dates[1:]
    else:
        return dates


@dataclasses.dataclass
class FV3Log:
    dates: List[datetime]
    totals: Mapping[str, List[float]]
    ranges: Mapping[str, List[Tuple[float, float]]]


def loads(log: str) -> FV3Log:
    """Parse the statistics information from the logs for an FV3 simulation

    Does not parse python-related diagnostics. It extracts the statistical
    information from blocks with the following format::

                2016           8           1           1           0           0
        ZS      6849.180      -412.0000       231.8707
        PS max =    1052.038      min =    439.9279
        Mean specific humidity (mg/kg) above 75 mb=   3.993258
        Total surface pressure (mb) =    985.9462
        mean dry surface pressure =    983.2382
        Total Water Vapor (kg/m**2) =   27.51812
        --- Micro Phys water substances (kg/m**2) --
        Total cloud water=  3.4418099E-0
        Total rain  water=  8.9859152E-0
        Total cloud ice  =  3.6711354E-0
        Total snow       =  1.4170707E-0
        Total graupel    =  1.3332913E-0
        --------------------------------------------
        TE ( Joule/m^2 * E9) =   2.633417
        UA_top max =    138.3022      min =   -45.65562
        UA max =    138.3022      min =   -52.00678
        VA max =    60.31665      min =   -67.41895
        W  max =    36.12219      min =   -11.35321
        Bottom w max =    5.591572      min =   -6.320749
        Bottom: w/dz max =   0.2113777      min =  -0.2430508
        DZ (m) max =   -21.97151      min =   -5737.403
        Bottom DZ (m) max =   -21.97151      min =   -32.53793
        TA max =    319.3932      min =    171.7615
        OM max =    79.11412      min =   -91.14186
        ZTOP      40.79998       34.23963       39.11939
        SLP max =    1064.040      min =    953.2686
        ATL SLP max =    1022.554      min =    1010.739
        fv_GFS Z500   5700.141       5790.903       5459.689       5864.878
        Cloud_top_P (mb) max =   1.0000000E+10  min =    14.69188
        Surf_wind_speed max =    33.00807      min =   4.1885415E-0
        T_bot:      319.3932       216.0579       289.6649
        ZTOP max =    40.79998      min =    34.23963
        W5km max =    13.85566      min =   -5.952400
        max reflectivity =    64.90421      dBZ
        sphum max =   2.3002494E-02  min =   9.9999342E-10
        liq_wat max =   2.2458597E-03  min =  -1.3747283E-19
        rainwat max =   8.2634874E-03  min =  -1.4335402E-18
        ice_wat max =   5.2885562E-03  min =  -5.2392893E-20
        snowwat max =   8.4350556E-03  min =  -9.2761843E-19
        graupel max =   1.3989445E-02  min =  -4.0843832E-05
        o3mr max =   1.7852419E-05  min =   2.2205915E-09
        sgs_tke max =    130.6761      min =   9.8705533E-10

    Currently it only supports the min/max and total statistics.

    Returns:
        fv3log: an object with the parsed statistical information
        
    
    """
    lines = log.splitlines()
    min_max_regex = re.compile(r"(.*)max *= *([0-9E\-\.]+) *min = *([0-9E\-\+\.]+)")
    zs_regexp = re.compile(r"ZS *([0-9E\-\.]+) +([0-9E\-\.]+) +([0-9E\-\+\.]+)")
    totals = defaultdict(list)
    min_max = defaultdict(list)
    dates = []

    mean_patterns = {
        "total surface pressure": parse_pattern(r"Total surface pressure \(mb\)"),
        "total water vapor": parse_pattern(r"Total Water Vapor \(kg/m\*\*2\)"),
        "mean dry surface pressure": parse_pattern("mean dry surface pressure"),
        "total cloud water": parse_pattern("Total cloud water"),
        "total rain water": parse_pattern(r"Total rain\s+water"),
        "total cloud ice": parse_pattern("Total cloud ice"),
        "total snow": parse_pattern("Total snow"),
        "total graupel": parse_pattern("Total graupel"),
    }

    date = None
    for line in lines:

        if "fv_restart_end" in line:
            # the logs end with a block like this::
            #
            # fv_restart_end u    =    5404162460055465614
            # ...
            #  ZS   6849.180      -412.0000       231.8706
            #
            # This break avoids appending the final date twice.
            break

        try:
            date = parse_date_stats_block(line)
        except (TypeError, ValueError):
            pass

        match = zs_regexp.search(line)
        if match:
            # The statistics block begins with the date followed by ZS::
            #         2016           8           2           8           0           0
            #  ZS      6849.180      -412.0000       231.8707
            dates.append(date)

        match = min_max_regex.search(line)
        if match:
            name, max_, min_ = match.groups()
            name = name.strip()
            min_max[name].append((float(min_), float(max_)))

        for variable, parser in mean_patterns.items():
            val = parser(line)
            if val is not None:
                totals[variable].append(val)

    return FV3Log(set_initial_date(dates), dict(totals), dict(min_max))
