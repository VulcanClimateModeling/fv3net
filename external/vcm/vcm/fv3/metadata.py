import xarray as xr
from typing import Mapping, Set, Callable, Sequence
import logging
from ..convenience import round_time
from ..safe import warn_if_intersecting

logger = logging.getLogger(__name__)


DIM_RENAME_INVERSE_MAP = {
    "x": {"grid_xt", "grid_xt_coarse"},
    "y": {"grid_yt", "grid_yt_coarse"},
    "tile": {"rank"},
    "x_interface": {"grid_x", "grid_x_coarse"},
    "y_interface": {"grid_y", "grid_y_coarse"},
}
VARNAME_SUFFIX_TO_REMOVE = ["_coarse"]
TIME_DIM_NAME = "time"
STANDARD_TO_GFDL_DIM_MAP = {
    "x_interface": "grid_x",
    "y_interface": "grid_y",
    "x": "grid_xt",
    "y": "grid_yt",
    "z": "pfull",
}


def standardize_fv3_diagnostics(
    ds: xr.Dataset, time: str = TIME_DIM_NAME
) -> xr.Dataset:
    """Standardize dimensions, coordinates, and attributes of FV3 diagnostic output
    
    Args:
        ds (xr.Dataset):
            Dataset of FV3GFS or SHiELD diagnostics outputs (either Fortran or
            python-derived), presumably opened from disk via zarr.
        time (str, optional):
            Name of the time coordinate dimension in the dataset
        
    Returns:
        The dataset with coordinate names and tile ranges set to CF "standard" values,
        variable name suffixes (e.g., "_coarse") removed, and data variable attributes
        ("long_name" and "units") set. The time coordinate is rounded to the nearest
        second.
    
    """

    if time in ds.coords:
        ds[time].attrs["calendar"] = "julian"

    funcs: Sequence[Callable[[xr.Dataset], xr.Dataset]] = [
        xr.decode_cf,
        _adjust_tile_range,
        _rename_dims,
        _round_time_coord,
        _remove_name_suffix,
        _set_missing_attrs,
    ]

    for func in funcs:
        ds = func(ds)

    return ds


def _adjust_tile_range(ds: xr.Dataset) -> xr.Dataset:

    if "tile" in ds:
        tiles = ds.tile

        if tiles.isel(tile=-1) == 6:
            ds = ds.assign_coords({"tile": tiles - 1})

    return ds


def _rename_dims(
    ds: xr.Dataset, rename_inverse: Mapping[str, Set[str]] = DIM_RENAME_INVERSE_MAP
) -> xr.Dataset:

    varname_target_registry = {}
    for target_name, source_names in rename_inverse.items():
        varname_target_registry.update({name: target_name for name in source_names})

    vars_to_rename = {
        var: varname_target_registry[str(var)]
        for var in ds.dims
        if var in varname_target_registry
    }
    ds = ds.rename(vars_to_rename)
    return ds


def _round_time_coord(ds: xr.Dataset, time_coord: str = TIME_DIM_NAME) -> xr.Dataset:

    if time_coord in ds.coords:
        new_times = round_time(ds[time_coord])
        ds = ds.assign_coords({time_coord: new_times})
    else:
        logger.debug(
            "Round time operation called on dataset missing a time coordinate."
        )

    return ds


def _set_missing_attrs(ds: xr.Dataset) -> xr.Dataset:

    for var in ds:
        da = ds[var]

        # True for some prognostic zarrs
        if "description" in da.attrs and "long_name" not in da.attrs:
            da.attrs["long_name"] = da.attrs["description"]

        if "long_name" not in da.attrs:
            da.attrs["long_name"] = var

        if "units" not in da.attrs:
            da.attrs["units"] = "unspecified"
    return ds


def _remove_name_suffix(
    ds: xr.Dataset, suffixes: Sequence[str] = VARNAME_SUFFIX_TO_REMOVE
) -> xr.Dataset:
    for target in suffixes:
        replace_names = {
            vname: str(vname).replace(target, "")
            for vname in ds.data_vars
            if target in str(vname)
        }

        warn_if_intersecting(ds.data_vars.keys(), replace_names.values())
        ds = ds.rename(replace_names)
    return ds


def gfdl_to_standard(ds: xr.Dataset):
    """Convert from GFDL dimension names (grid_xt, etc) to "standard"
    names (x, y, z)
    """

    key, val = STANDARD_TO_GFDL_DIM_MAP.keys(), STANDARD_TO_GFDL_DIM_MAP.values()
    inverse = dict(zip(val, key))

    return ds.rename({key: val for key, val in inverse.items() if key in ds.dims})


def standard_to_gfdl(ds: xr.Dataset):
    """Convert from "standard" names to GFDL names"""
    return ds.rename(
        {key: val for key, val in STANDARD_TO_GFDL_DIM_MAP.items() if key in ds.dims}
    )
