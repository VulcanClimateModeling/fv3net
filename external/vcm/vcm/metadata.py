rename = {
    "x_interface": "grid_x",
    "y_interface": "grid_y",
    "x": "grid_xt",
    "y": "grid_yt",
    "z": "pfull",
}


def gfdl_to_standard(ds):
    """Convert from GFDL dimension names (grid_xt, etc) to standard
    names (x, y, z)
    """

    key, val = rename.keys(), rename.values()
    inverse = dict(zip(val, key))

    return ds.rename({key: val for key, val in inverse.items() if key in ds.dims})


def standard_to_gfdl(ds):
    """Convert from standard names to GFDL names"""
    return ds.rename({key: val for key, val in rename.items() if key in ds.dims})
