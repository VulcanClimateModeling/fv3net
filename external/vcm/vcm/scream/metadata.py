import xarray as xr
import logging
import numpy as np
from typing import Callable, Sequence

TIME_DIM_NAME = "time"

logger = logging.getLogger(__name__)


def _mask_invalid_value(ds: xr.Dataset) -> xr.Dataset:
    ds = xr.where(ds > 1e30, np.nan, ds)
    return ds


def _set_missing_attrs(ds: xr.Dataset) -> xr.Dataset:

    for var in ds:
        da = ds[var]

        # True for some prognostic zarrs
        if "description" in da.attrs and "long_name" not in da.attrs:
            da.attrs["long_name"] = da.attrs["description"]

        if "long_name" not in da.attrs:
            da.attrs["long_name"] = var

        if "units" not in da.attrs:
            da.attrs["units"] = "unspecified"
    return ds


def standardize_scream_diagnostics(
    ds: xr.Dataset, time: str = TIME_DIM_NAME
) -> xr.Dataset:
    """Standardize SCREAM diagnostics to a common format.

    Args:
        ds (xr.Dataset):
            Dataset of scream diagnostics outputs,
            presumably opened from disk via zarr.
        time (str, optional):
            Name of the time coordinate dimension in the dataset

    Returns:
        Data variable attributes ("long_name" and "units") are set.
        Clean up invalid values that hasn't been fixed upstream.
    """
    if time in ds.coords:
        ds[time].attrs["calendar"] = "julian"
    funcs: Sequence[Callable[[xr.Dataset], xr.Dataset]] = [
        xr.decode_cf,
        _set_missing_attrs,
        _mask_invalid_value,
    ]
    for func in funcs:
        ds = func(ds)
    return ds
