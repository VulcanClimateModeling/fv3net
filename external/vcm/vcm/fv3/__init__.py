from . import logs
from .metadata import standardize_fv3_diagnostics, gfdl_to_standard, standard_to_gfdl
