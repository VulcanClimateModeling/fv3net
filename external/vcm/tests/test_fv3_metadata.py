import pytest
import xarray as xr
import numpy as np
import cftime
from numpy.random import randint
import numpy as np

from vcm.fv3.metadata import (
    _adjust_tile_range,
    _rename_dims,
    _set_missing_attrs,
    standardize_fv3_diagnostics,
    gfdl_to_standard,
    standard_to_gfdl
)


@pytest.mark.parametrize(
    "ds",
    [
        xr.Dataset(coords={"tile": np.arange(1, 7)}),
        xr.Dataset(coords={"tile": np.arange(6)}),
    ],
)
def test__check_tile_range(ds):

    expected = np.arange(6)
    expected_da = xr.DataArray(expected, dims=["tile"], coords={"tile": expected})
    tile_result = _adjust_tile_range(ds).tile
    xr.testing.assert_equal(tile_result, expected_da)


def _create_dataset(*dims, with_coords=True):
    if with_coords:
        coords = {dim: np.arange(i + 1) for i, dim in enumerate(dims)}
        ds = xr.Dataset(coords=coords)
    else:
        arr = np.zeros([i + 1 for i in range(len(dims))])
        da = xr.DataArray(arr, dims=dims)
        ds = xr.Dataset({"varname": da})
    return ds


@pytest.mark.parametrize(
    "input_dims, rename_inverse, renamed_dims",
    [
        ({"x", "y"}, {}, {"x", "y"}),
        ({"x", "y"}, {"y_out": {"y"}}, {"x", "y_out"}),
        ({"x", "y"}, {"y_out": {"y", "y2"}}, {"x", "y_out"}),
        ({"x", "y"}, {"x_out": {"x"}, "y_out": {"y", "y2"}}, {"x_out", "y_out"}),
        ({"x", "y"}, {"z_out": {"z"}}, {"x", "y"}),
    ],
)
def test__rename_dims(input_dims, rename_inverse, renamed_dims):
    # datasets can have dimensions with or without coordinates, so cover both cases
    for with_coords in [True, False]:
        ds_in = _create_dataset(*input_dims, with_coords=with_coords)
        ds_out = _rename_dims(ds_in, rename_inverse=rename_inverse)
        assert set(ds_out.dims) == renamed_dims


@pytest.fixture
def xr_darray():
    data = np.arange(16).reshape(4, 4)
    x = np.arange(4)
    y = np.arange(4)

    da = xr.DataArray(data, coords={"x": x, "y": y}, dims=["x", "y"],)

    return da


@pytest.mark.parametrize(
    "attrs",
    [
        {},
        {"units": "best units"},
        {"long_name": "name is long!"},
        {"units": "trees", "long_name": "number of U.S. trees"},
    ],
)
def test__set_missing_attrs(attrs, xr_darray):

    xr_darray.attrs.update(attrs)
    res = _set_missing_attrs(xr_darray.to_dataset(name="data"))
    assert "long_name" in res.data.attrs
    assert "units" in res.data.attrs


def test__set_missing_attrs_description(xr_darray):

    attrs = {"description": "a description will be converted to a longname"}
    xr_darray.attrs.update(attrs)
    res = _set_missing_attrs(xr_darray.to_dataset(name="data"))
    assert res.data.attrs["long_name"] == attrs["description"]


DIM_NAMES = ["grid_xt", "grid_yt", "tile"]
DATA_VARS = ["a", "b"]
TIME_DIM = 10


def _create_raw_dataset():
    coords = {dim: np.arange(i + 1) for i, dim in enumerate(DIM_NAMES)}
    sizes = {dim: len(coords[dim]) for dim in DIM_NAMES}
    dataset = {}
    for data_var in DATA_VARS:
        arr = np.ones([*sizes.values()])
        dataset[data_var] = xr.DataArray(arr, dims=DIM_NAMES)
    ds = xr.Dataset(dataset, coords=coords)
    time_coord = [
        cftime.DatetimeJulian(2016, 1, n + 1, 0, 0, 0, randint(100))
        for n in range(TIME_DIM)
    ]
    ds["time"] = time_coord
    return ds


def test_standardize_fv3_diagnostics(tmpdir, regtest):
    ds = _create_raw_dataset()
    ds.to_zarr(str(tmpdir.join("fv3_diag.zarr")), consolidated=True)
    diag = xr.open_zarr(str(tmpdir.join("fv3_diag.zarr")), consolidated=True).load()
    standardized_ds = standardize_fv3_diagnostics(diag)
    with regtest:
        print(standardized_ds)
        print(standardized_ds.time)
        print(standardized_ds.time.attrs)


def test_gfdl_to_standard_dims_correct():
    data = xr.Dataset(
        {
            "2d": (["tile", "grid_yt", "grid_xt"], np.ones((1, 1, 1))),
            "3d": (["tile", "pfull", "grid_yt", "grid_xt"], np.ones((1, 1, 1, 1))),
        }
    )

    ans = gfdl_to_standard(data)
    assert set(ans.dims) == {"tile", "x", "y", "z"}


def test_gfdl_to_standard_is_inverse_of_standard_to_gfdl():

    data = xr.Dataset(
        {
            "2d": (["tile", "grid_yt", "grid_xt"], np.ones((1, 1, 1))),
            "3d": (["tile", "pfull", "grid_yt", "grid_xt"], np.ones((1, 1, 1, 1))),
        }
    )

    back = standard_to_gfdl(gfdl_to_standard(data))
    xr.testing.assert_equal(data, back)
