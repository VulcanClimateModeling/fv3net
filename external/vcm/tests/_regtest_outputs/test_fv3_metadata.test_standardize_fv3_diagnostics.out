<xarray.Dataset>
Dimensions:  (tile: 3, time: 10, x: 1, y: 2)
Coordinates:
  * x        (x) int64 0
  * y        (y) int64 0 1
  * tile     (tile) int64 0 1 2
  * time     (time) object 2016-01-01 00:00:00 ... 2016-01-10 00:00:00
Data variables:
    a        (x, y, tile) float64 ...
    b        (x, y, tile) float64 ...
<xarray.DataArray 'time' (time: 10)>
array([cftime.DatetimeJulian(2016, 1, 1, 0, 0, 0, 0),
       cftime.DatetimeJulian(2016, 1, 2, 0, 0, 0, 0),
       cftime.DatetimeJulian(2016, 1, 3, 0, 0, 0, 0),
       cftime.DatetimeJulian(2016, 1, 4, 0, 0, 0, 0),
       cftime.DatetimeJulian(2016, 1, 5, 0, 0, 0, 0),
       cftime.DatetimeJulian(2016, 1, 6, 0, 0, 0, 0),
       cftime.DatetimeJulian(2016, 1, 7, 0, 0, 0, 0),
       cftime.DatetimeJulian(2016, 1, 8, 0, 0, 0, 0),
       cftime.DatetimeJulian(2016, 1, 9, 0, 0, 0, 0),
       cftime.DatetimeJulian(2016, 1, 10, 0, 0, 0, 0)], dtype=object)
Coordinates:
  * time     (time) object 2016-01-01 00:00:00 ... 2016-01-10 00:00:00
{}
